### PR TITLE
Add signature for attr.evolve

### LINKED ISSUE #14525: attrs plugin should support attrs.evolve
**Feature**

In the attrs plugin, we should add special treatment for [attrs.evolve](https://www.attrs.org/en/stable/api.html#attrs.evolve).

**Pitch**

We should validate attr names and types match.

### PR DESCRIPTION
**Pull request by ikonst**
Validate `attr.evolve` calls to specify correct arguments and types.

The implementation makes it so that at every point where `attr.evolve` is called, the signature is modified to expect the attrs class' initializer's arguments (but so that they're all kw-only and optional).

Notes:
- Added `class dict: pass` to some fixtures files since our attrs type stubs now have **kwargs and that triggers a `builtin.dict` lookup in dozens of attrs tests.
- Looking up the type of the 1st argument with `ctx.api.expr_checker.accept(inst_arg)` which is a hack since it's not part of the plugin API. This is a compromise for due to #10216.

Fixes #14525.

### CODE REVIEW

File: mypy/plugins/default.py [hunk 1]
```diff
@@ -45,6 +46,15 @@ def get_function_hook(self, fullname: str) -> Callable[[FunctionContext], Type]
             return singledispatch.create_singledispatch_function_callback
         return None
 
+    def get_function_signature_hook(
+        self, fullname: str
+    ) -> Callable[[FunctionSigContext], FunctionLike] | None:
+        from mypy.plugins import attrs
+
+        if fullname in ("attr.evolve", "attr.assoc"):
```
Comment (TeamSpen210) [inline mypy/plugins/default.py:54]: There’s also `attrs.evolve()` - maybe that’s just an alias and will work, but it should be tested. Perhaps these should be moved to a constant with the others at the start of the attrs plug-in module also?

Comment (ikonst) [reply mypy/plugins/default.py:54]: Yep, yep. All still WIP.

**Comment by ikonst** [comment]
mypy-primer [an error in home-assistant](https://github.com/home-assistant/core/blob/e9543a72547210c84e4c17ed45774f3548bb5dc5/homeassistant/helpers/area_registry.py#L219):
```python
new = self.areas[area_id] = attr.evolve(old, **new_values)
```

Typing `new_values` as a `TypedDict` can address that. The misleading error is mentioned in https://github.com/python/mypy/issues/14073.

File: mypy/plugins/default.py (hunk 1 above)
Comment (ikonst) [reply mypy/plugins/default.py]: Ready now!

File: test-data/unit/fine-grained.test [hunk 2]
```diff
@@ -1809,8 +1809,8 @@ def f() -> Iterator[None]:
 [typing fixtures/typing-medium.pyi]
 [builtins fixtures/list.pyi]
 [triggered]
-2: <b>, __main__
-3: <b>, __main__, a
+2: <b>, <b[wildcard]>, __main__
+3: <b>, <b[wildcard]>, __main__, a
```
Comment (ikonst) [inline test-data/unit/fine-grained.test:1813]: So.... the addition of `dict` in fixtures/list.pyi now causes `__annotations__` to exist:
https://github.com/python/mypy/blob/bac9e77eaeb36c0535bc05cb1faf6eced25b8af1/mypy/semanal.py#L637-L639

which then makes it one of the names that's part of the diff:
https://github.com/python/mypy/blob/a08388cf6de053f659a9d663387f8f71e68664d8/mypy/server/update.py#L781-L795
(sounds reasonable to me)

Generally I noticed 3 approaches when a built-in is missing:
- crash — like when handling **kwargs here:
   https://github.com/python/mypy/blob/e778a58066f23982d5cbe1df5317d6540c5902fd/mypy/types.py#L3607
- avoid doing a thing — like when not adding `__annotations__` above
- fallback to `object` — like when deciding how to type `NamedTuple._field_types`:
   https://github.com/python/mypy/blob/dcf910e05fdcff00b1fb9a8b9ca85f9a6667b946/mypy/semanal_namedtuple.py#L484-L486

Granted it's a "testing problem" since otherwise those builtins can be taken for granted, but would be good to have a shared/recommended approach to make tests maintainable (would rather not have to do whack-a-mole in unrelated tests)... cc @97littleleaf11 @hauntsaninja

**Comment by ikonst** [comment]
@AlexWaygood can label with attrs? :)

**Comment by ikonst** [comment]
🤔  Curious why the mypy_primer diff here is gone:
https://github.com/home-assistant/core/blob/8968ed1c47314db4b9ee7eeb302c92c4e71915a3/homeassistant/helpers/area_registry.py#L219

**Comment by ikonst** [comment]
@JelleZijlstra would love to get some eyes on this (I was waiting for #14575 to merge to reduce the diff)

**Comment by JelleZijlstra** [comment]
I can take a look in the next few days.

This function is basically equivalent to `dataclasses.replace`, right? Would be nice if we can support that function in a similar way (in a separate PR).

File: mypy/types.py [hunk 3]
```diff
@@ -1775,7 +1775,7 @@ def copy_modified(
         self: CT,
         arg_types: Bogus[Sequence[Type]] = _dummy,
         arg_kinds: Bogus[list[ArgKind]] = _dummy,
-        arg_names: Bogus[list[str | None]] = _dummy,
+        arg_names: Bogus[Sequence[str | None]] = _dummy,
```
Comment (JelleZijlstra) [inline mypy/types.py:1778]: What is the reason for this change? I'd prefer to avoid touching the mypy core in this PR.

**Review by JelleZijlstra — APPROVED** [review]
Looks good, just one minor point.

File: mypy/types.py (hunk 3 above)
Comment (ikonst) [reply mypy/types.py:1778]: Because of list's invariance.  In `CallableType.__init__`, we copy `self.arg_names = list(arg_names)` so we can safely make it a `Sequence[str | None]`.

Otherwise, this fails since it passes `list[str]`:
```python
arg_names=["inst"] + attrs_init_type.arg_names[1:],
```

Comment (ikonst) [reply mypy/types.py:1778]: Broke out into #14840 — can merge then update this branch.

File: mypy/plugins/attrs.py [hunk 4]
```diff
@@ -888,3 +891,64 @@ def add_method(
...
+
+    if len(ctx.args[0]) != 1:
+        return ctx.default_signature  # leave it to the type checker to complain
+
+    inst_arg = ctx.args[0][0]
+
+    # <hack>
+    assert isinstance(ctx.api, TypeChecker)
+    inst_type = ctx.api.expr_checker.accept(inst_arg)
+    # </hack>
```
Comment (ikonst) [inline mypy/plugins/attrs.py:930]: @JelleZijlstra now that we got this merged 😅 what would be the right place to discuss whether this should be promoted to be formal plugin API?

Comment (JelleZijlstra) [reply mypy/plugins/attrs.py:930]: Open an issue, I suppose.

Comment (ikonst) [reply mypy/plugins/attrs.py:930]: 👉  #14845

**Comment by LukasKrocek** [comment]
hey, I have tried to run this in my project, doen't work well when evolving generic classes

`@frozen
class A(Generic[T]):
    field: T

a = A("1")

evolve(a, field="2")`

gives 2 errors:

Argument 1 to "evolve" of "A[str]" has incompatible type "A[str]"; expected "A[T]"  [arg-type]
Argument "field" to "evolve" of "A[str]" has incompatible type "int"; expected "T"  [arg-type]

**Comment by ikonst** [comment]
@LukasKrocek this was just mentioned in the review for #14849 — addressed in c0058950753ef275edb6abd6775660a4b0dc971e, but here it would actually be simpler

**Comment by ikonst** [comment]
Generics should be addressed in #15016.
